### DISCUSSION TITLE
CloudEvents

### DISCUSSION THREAD

**Discussion by tfahringer** [alexzhornyak/SCXML-tutorial#26] [Q&A]
CloudEvents is a major standard for events in the cloud-edge continuum. How could I connect CloudEvents with the generic events of SCXML.
Basically I want to use CloudEvents as the underlying event system for invoking and transferring events.
How can I interface to CloudEvents from SCXML?

**Comment by alexzhornyak** [comment]
#### There are [2 event processors that are described in SCXML standard](https://www.w3.org/TR/scxml/#SCXMLEventProcessor):
- **SCXML Event I/O Processor** - The SCXML Event I/O Processor is intended to transport messages between SCXML sessions
- **Basic HTTP Event I/O Processor** - The Basic HTTP Event I/O Processor is intended as a minimal interoperable mechanism for sending and receiving events to and from external components and SCXML 1.0 implementations. Support for the Basic HTTP Event I/O Processor is optional.

> **NOTE:** If your **SCXML framework (uscxml, Qt scxml, scion etc.)** allows you to register **custom event processor** you can do it according to framework documentation

#### Example of custom ZeroMQ based event processor overrided from uscxml framework SCXMLIOProcessor class

```cpp
class ZmqIOProcessor : public SCXMLIOProcessor {
	ScxmlTask *m_ScxmlTask;
public:
	ZmqIOProcessor(IOProcessorCallbacks* callbacks, ScxmlTask *AScxmlTask);

	virtual std::list<std::string> getNames() override {
		std::list<std::string> names;
		names.push_back(protocol());
		return names;
	};

	virtual std::shared_ptr<IOProcessorImpl> create(IOProcessorCallbacks* callbacks) override;

	virtual void eventFromSCXML(const std::string& target, const Event& event) override {
		// sending received Event from state machine by ZeroMQ network service
	}

	static std::string protocol() { return "zmq_protocol"; }
};
```

#### [Other uscxml framwork ioprocessor examples](https://github.com/tklab-tud/uscxml/tree/master/src/uscxml/plugins/ioprocessor)
